pkg/utils/symbol_table.go

```diff
@@ -0,0 +1,212 @@
+package utils
+
+import (
+	"errors"
+	"sort"
+	"sync"
+)
+
+// The Symbol interface defines what is needed from a symbol implementation in
+// order to facilitate the lookup functionalities provided by SymbolTable.
+// Implementations of Symbol can hold various types of information relevant to
+// the type of symbol they represent.
+type Symbol interface {
+	Name() string
+	Address() uint64
+	Contains(address uint64) bool
+}
+
+// SymbolTable is used to hold information about symbols (mapping from symbolic
+// names used in code to their address) in a certain executable.
+// It can be used to hold symbols from an ELF binary, or symbols of the entire
+// kernel and its modules.
+// It provides functions to lookup symbols by address and name.
+type SymbolTable[T Symbol] struct {
+	mu sync.RWMutex
+	// All symbols sorted by their address in descending order,
+	// for quick binary searches by address.
+	sortedSymbols []*T
```

Comment:
Why `*T`? `Symbol` is already a pointer by definition.

Reply:
Are you sure it is? From my understanding if I use `T` then `sortedSymbols` contains all of the symbols inline, which I tried but it made it harder to manage symbol references in the name to symbol map.

Reply:
IIRC say you have a struct `S` which implements some interface `I`. Then you can pass `s S` as `I` only with `&s`.

Reply:
So what you're doing is requiring concretely a `**KernelSymbol` instead of a `*KernelSymbol` as `Symbol`.

Reply:
> IIRC say you have a struct `S` which implements some interface `I`. Then you can pass `s S` as `I` only with `&s`.

If I understand correctly, this applies only if the struct methods are implemented using a pointer receiver, notice how I implemented them without it and given a non-pointer type when creating the underlying symbol table.

Reply:
Pointer receivers tend to be more efficient but that's besides the point. Yes you're correct but then why add the indirection at this point? Just stick to the (hopefully) stack allocated objects. Now you're forcing heap escape. If we're going through pointer eventually might as well just author the struct as a pointer receiver and make the code simpler.

Reply:
Not sure I understand what you're referring to with stack allocated objects. The reason I chose to use `*T` for `sortedSymbols` and `symbolsByName` is to make sure we're dealing with pointers, otherwise a future user of `SymbolTable` could use a non-pointer type and all of the symbols in `symbolsByName` will be duplicated instead of using their pointers. I guess this also means that if I use `*T` then a future user could use a pointer type which will result in double indirection.

If you think using `T` and passing `*KernelSymbol` as the symbol type is preferrable I'll change it, let me know.